### PR TITLE
Streaming SFT support

### PR DESCRIPTION
**Pull request by djsaunde**
# Description

Streaming datasets should not be limited to pretraining only. This PR changes that.

Also removed `--iterable` preprocess CLI arg since it was confusing / maybe not advisable to do.

This is a follow-up to #3087 which is smaller in scope since things were broken / getting complicated in that PR.

TODO:
- [x] `preprocess --iterable` deprecation (from #3087)
- [x] rework `sample_packing` / `pretrain_multipack_attn` oddness
  - the latter should be always enabled in SFT setting IMO
- [x] brief docs + sample config

Follow-ups:
- Add dataset mixing strategies (see #3087)
- Add offline RL streaming support
- Deprecate `pretraining_dataset` config
- Remove `--iterable` CLI arg
- Refactor
  - Code is hard to follow
  - Unmix pretraining and streaming logic

## Motivation and Context

Some folks have been asking for better dataset streaming support. This is part of a larger effort to refactor / improve the logic around how we handle datasets / dataloading.

## How has this been tested?

- [x] Unit tests
- [x] Smoke tests
- [x] Manually (comparing average loss, grad norm, ...)
  - stream vs. no stream, packing
  - stream vs. no stream, no packing

## Summary by CodeRabbit

- **New Features**
  - Streaming datasets for training: sample packing, attention-isolation controls, automatic streaming when a pretraining dataset is provided, new streaming config flags (streaming, streaming_multipack_buffer_size).

- **Documentation**
  - Added “Streaming Datasets” guide and examples/streaming README with pretrain and SFT example configs.

- **Bug Fixes**
  - Preserve streaming behavior in fallback loading, safer long-sequence handling, and reduced repeated empty-prompt warnings.

- **Chores**
  - Deprecated preprocess --iterable; use streaming: true or --streaming.

- **Tests**
  - New unit and end-to-end tests covering streaming behavior.

### CODE REVIEW

File: src/axolotl/utils/data/pretraining.py [hunk 1]
```diff
@@ -176,45 +176,51 @@ def encode_pretraining(
...
     else:
         encode = functools.partial(
-            encode_pretraining,
+            encode_streaming,
             tokenizer,
-            max_tokens,
+            max_tokens=cfg.sequence_len,
             text_column=cfg.pretraining_dataset[0].text_column or "text",
             concatenate=cfg.pretraining_sample_concatenation is True,
         )
```
[the remark this answers is not in the document]
Comment (djsaunde) [reply src/axolotl/utils/data/pretraining.py]: This is not applicable since this is unreachable by anything other than pretraining without sample packing.

File: docs/streaming.qmd [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+Streaming enables memory-efficient training with large datasets by loading data
+incrementally rather than loading the entire dataset into memory at once.
+
+Use streaming when:
+
+- Your dataset is too large to fit in memory
+  - E.g., you're doing pretraining with massive text corpora
+- You want to start training immediately without preprocessing the entire dataset
+
+Streaming works with both remote and locally stored datasets!
```
Comment (winglian) [inline docs/streaming.qmd:15]: We should add a callout here about only supporting single dataset configuration for streaming for now.

File: src/axolotl/utils/data/sft.py [hunk 3]
```diff
@@ -128,22 +122,46 @@ def _load_datasets():
...
+        dataset_config = _extract_pretraining_config(cfg)
+        train_dataset = _load_streaming_dataset(dataset_config, cfg, tokenizer)
+    else:
+        if cfg.sample_packing:
+            # TODO(djsaunde): Implement for multiple datasets
+            if len(cfg.datasets) > 1:
+                raise NotImplementedError(
+                    "Sample packing with multiple streaming datasets is not yet "
+                    "supported"
+                )
```
Comment (winglian) [inline src/axolotl/utils/data/sft.py]: Could we move this to the schema validation?

File: docs/streaming.qmd (hunk 2 above)
Comment (djsaunde) [reply docs/streaming.qmd:15]: Can do. I'd like to add multidataset support ASAP in a followup

File: src/axolotl/cli/preprocess.py [hunk 4]
```diff
@@ -35,10 +35,20 @@ def do_preprocess(cfg: DictDefault, cli_args: PreprocessCliArgs) -> None:
...
     check_user_token()
 
+    if cli_args.iterable:
+        LOG.error(
+            "The --iterable CLI argument for 'axolotl preprocess' is no longer "
+            "supported. For training, set 'streaming: true' in your YAML config or "
+            "pass '--streaming' in your 'axolotl train' command for on-the-fly "
+            "preprocessing."
+        )
+        return
```
Comment (NanoCode012) [inline src/axolotl/cli/preprocess.py:45]: Is it possible to put this in our pydantic validation or does it not cover cli args?

File: docs/streaming.qmd [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+### Pretraining with Streaming
+
+For pretraining tasks, streaming is automatically enabled when using `pretraining_dataset`:
+
+```yaml
+pretraining_dataset:
+  - path: HuggingFaceFW/fineweb-edu
+    type: pretrain
+    text_column: text
+    split: train
```
Comment (NanoCode012) [inline docs/streaming.qmd:41]: Do you think it's better to just deprecate `pretraining_dataset` to use `type: completion` with streaming instead?

File: docs/streaming.qmd [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+# Optionally, enable sample packing
+streaming_multipack_buffer_size: 10000
+sample_packing: true
+```
+
+### SFT with Streaming
+
+For supervised fine-tuning with streaming:
```
Comment (NanoCode012) [inline docs/streaming.qmd:50]: We should also add this to `docs/dataset-formats/index.qmd`. I had some points where I said streaming was only available for pretraining.

File: src/axolotl/utils/data/sft.py [hunk 7]
```diff
@@ -128,22 +122,41 @@ def _load_datasets():
...
+            train_dataset = _load_streaming_dataset(dataset_config, cfg, tokenizer)
+        else:
+            # Use legacy loading function for non-packed streaming datasets
+            train_dataset, eval_dataset, prompters = _load_and_prepare_datasets(
+                tokenizer,
+                cfg,
+                split="train",
+                processor=processor,
+                streaming=True,
+            )
```
Comment (NanoCode012) [inline src/axolotl/utils/data/sft.py]: nit: maybe we can refactor the first `else` to `elif cfg.sample_packing` to indent left one block

File: src/axolotl/utils/data/sft.py [hunk 8]
```diff
@@ -153,14 +166,15 @@ def _prepare_pretraining_dataset(
...
 
     if cfg.dataset_exact_deduplication:
-        LOG.info("Deduplication not available for pretrained datasets")
+        LOG.info("Deduplication not available for streaming datasets")
 
-    # For pretraining, we return max_steps directly from config
-    return train_dataset, eval_dataset, cfg.max_steps, []
+    # For streaming, we return max_steps directly from config or -1 if not set
+    total_num_steps = cfg.max_steps if cfg.max_steps else -1
+    return train_dataset, eval_dataset, total_num_steps, []
```
Comment (NanoCode012) [inline src/axolotl/utils/data/sft.py:173]: Do we need to mention this is necessary on the `streaming` docs? I recalled that this was needed for `pretraining_datasets` otherwise we won't know when to end a while back.

File: src/axolotl/utils/schemas/config.py [hunk 9]
```diff
@@ -495,13 +489,23 @@ class AxolotlInputConfig(
...
     )
+    pretraining_sample_concatenation: bool | None = Field(
+        default=None,
+        json_schema_extra={
+            "description": "whether to concatenate samples during pretraining",
+        },
+    )
+
+    streaming: bool | None = None
+    streaming_multipack_buffer_size: int | None = 10_000
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/config.py]: We should add Field description for these three and `pretrain_multipack_buffer_size` too (with deprecation warning)

File: src/axolotl/utils/schemas/validation.py [hunk 10]
```diff
@@ -60,6 +60,20 @@ def check_dataset_or_pretraining_dataset(cls, data):
...
 
+    @model_validator(mode="before")
+    @classmethod
+    def check_pretraining_streaming_deprecation(cls, data):
+        # TODO(djsaunde): remove this check + implement change for 0.13.0 release
+        if data.get("pretraining_dataset") and not data.get("streaming"):
+            LOG.warning(
+                "Setting `pretraining_dataset` without explicitly setting `streaming: "
+                "true` is deprecated. In a future release, streaming will not be "
+                "automatically enabled when using pretraining_dataset. Please "
```
Comment (NanoCode012) [inline src/axolotl/utils/schemas/validation.py:71]: Are we planning to allow pretraining dataset without streaming?

**Review by NanoCode012 — COMMENTED** [review]
Nice work refactoring out `preprocess_iterable`.

File: docs/streaming.qmd [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+order: 10
+---
+
+Streaming enables memory-efficient training with large datasets by loading data
+incrementally rather than loading the entire dataset into memory at once.
+
+Use streaming when:
+
+- Your dataset is too large to fit in memory
+  - E.g., you're doing pretraining with massive text corpora
```
Comment (salmanmohammadi) [inline docs/streaming.qmd]:
```suggestion
- Your dataset is too large to fit in memory e.g. when you're doing pretraining with massive text corpora
```

File: docs/streaming.qmd [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+For more information, see our [documentation](multipack.qmd) on multipacking.
+
+## Important Considerations
+
+### Memory Usage
+
+While streaming reduces memory usage compared to loading entire datasets, you still need
+to consider:
+
+- The `streaming_multipack_buffer_size` affects memory usage
```
Comment (salmanmohammadi) [inline docs/streaming.qmd]:
```suggestion
- You can control the memory usage by adjusting `streaming_multipack_buffer_size` 
```

File: docs/streaming.qmd [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+- Sample packing requires buffering multiple samples
+- Shuffling requires additional memory for the shuffle buffer
+
+### Performance
+
+- Streaming may have slightly higher latency compared to preprocessed datasets
```
Comment (salmanmohammadi) [inline docs/streaming.qmd]:
```suggestion
- Streaming may have slightly higher latency compared to preprocessed datasets, as samples are processed on-the-fly
```

File: docs/streaming.qmd [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+- Network speed matters when streaming from remote sources
```
Comment (salmanmohammadi) [inline docs/streaming.qmd]:
```suggestion
- Network speed and disk read speed are important when streaming from remote sources or a local dataset, respectively
```

File: docs/streaming.qmd [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+- Consider using `axolotl preprocess` for smaller or more frequently used datasets
```
Comment (salmanmohammadi) [inline docs/streaming.qmd:108]: How does this work? I preprocess and stream? Or is this instead of?

File: src/axolotl/cli/args.py [hunk 16]
```diff
@@ -14,9 +14,13 @@ class PreprocessCliArgs:
     prompter: Optional[str] = field(default=None)
     download: Optional[bool] = field(default=True)
     iterable: Optional[bool] = field(
-        default=None,
+        default=False,
         metadata={
-            "help": "Use IterableDataset for streaming processing of large datasets"
+            "help": (
+                "[DEPRECATED] No longer supported. For streaming datasets, use "
```
Comment (salmanmohammadi) [inline src/axolotl/cli/args.py]: I was thinking earlier that we should try and be explicit about deprecation versions - mainly for our sake so we know when to go back and remove features

File: src/axolotl/utils/data/sft.py [hunk 17]
```diff
@@ -153,14 +166,15 @@ def _prepare_pretraining_dataset(
             cfg,
             split="test",
             processor=processor,
-            preprocess_iterable=preprocess_iterable,
+            streaming=False,
         )
 
     if cfg.dataset_exact_deduplication:
-        LOG.info("Deduplication not available for pretrained datasets")
+        LOG.info("Deduplication not available for streaming datasets")
```
Comment (salmanmohammadi) [inline src/axolotl/utils/data/sft.py]: Should we catch this in the schema validation instead? This seems like it could cause confusing results for users who don't closely watch the logs

File: src/axolotl/utils/data/streaming.py [hunk 18]
```diff
@@ -176,45 +176,57 @@ def encode_pretraining(
...
     ds_wrapper_fn,
-    max_tokens=2048,
-    batch_size=1,
-    seed=42,
-    buffer_size=10_000,
 ):
     if cfg.sample_packing:
+        # For SFT (non-pretraining) datasets, always use multipack_attn=True to ensure
+        # attention isolation between packed sequences
+        multipack_attn = (
```
Comment (salmanmohammadi) [inline src/axolotl/utils/data/streaming.py:188]: What are the implications for adding this new default behaviour?

File: src/axolotl/utils/data/streaming.py [hunk 19]
```diff
@@ -274,8 +285,6 @@ def encode_packed_pretraining(
     for batch in sampler:
         for data in batch:
             features = train_dataset[data]
-            if "num_truncated_tokens" in features:
```
Comment (salmanmohammadi) [inline src/axolotl/utils/data/streaming.py:277]: Was this being used anywhere before?

File: src/axolotl/utils/schemas/config.py [hunk 20]
```diff
@@ -495,13 +489,23 @@ class AxolotlInputConfig(
     pose_max_context_len: int | None = None
     pose_num_chunks: int | None = None
 
-    pretrain_multipack_buffer_size: int | None = 10_000
+    # Deprecated: Use streaming_multipack_buffer_size instead
+    pretrain_multipack_buffer_size: int | None = None
```
Comment (salmanmohammadi) [inline src/axolotl/utils/schemas/config.py]: You could add a `deprecated` kwarg in the Field constructor here

File: src/axolotl/utils/data/streaming.py (hunk 19 above)
Comment (NanoCode012) [reply src/axolotl/utils/data/streaming.py:277]: It was duplicate of the one below.

File: src/axolotl/cli/preprocess.py (hunk 4 above)
Comment (djsaunde) [reply src/axolotl/cli/preprocess.py:45]: Yeah, pydantic validation just covers the `AxolotlInputArgs` model.

File: docs/streaming.qmd (hunk 5 above)
Comment (djsaunde) [reply docs/streaming.qmd:41]: I like this idea, maybe we could add as a followup instead though? Might be too big of a breaking change for users of this config.

**Review by winglian — APPROVED** [review]
